metadata/src/test/java/org/apache/kafka/controller/LocalLogManagerTest.java

```diff
@@ -0,0 +1,258 @@
+            this.index = index;
+            this.message = message;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(epoch, index, message);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (!(o instanceof TestCommit)) {
+                return false;
+            }
+            TestCommit other = (TestCommit) o;
+            return epoch == other.epoch &&
+                index == other.index &&
+                message.equals(other.message);
+        }
+
+        @Override
+        public String toString() {
+            return "TestCommit(epoch=" + epoch + ", index=" + index +
+                ", message=" + message + ")";
+        }
+    }
+
+    static class CommitChecker implements CommitHandler {
+        private final List<TestCommit> commits;
+        private final Map<Integer, Integer> cursors = new HashMap<>();
+
+        CommitChecker(List<TestCommit> commits) {
+            this.commits = commits;
+        }
+
+        @Override
+        public synchronized void handle(int nodeId, long epoch, long index,
+                                        ApiMessage message) throws Exception {
+            int commitNumber = cursors.getOrDefault(nodeId, 0);
+            TestCommit newCommit = new TestCommit(epoch, index, message);
+            TestCommit expectedCommit = commits.get(commitNumber);
+            assertEquals("Unexpected commit on node " + nodeId, expectedCommit, newCommit);
+            cursors.put(nodeId, commitNumber + 1);
+        }
+
+        synchronized void verifyCursorsHaveReachedCommitNumber(int expectedNumCursors,
+                                                               int expectedCommitNumber) {
+            if (cursors.size() < expectedNumCursors) {
+                throw new RuntimeException("Expected " + expectedNumCursors +
+                    " cursors, but there were only " + cursors.size());
+            }
+            for (Map.Entry<Integer, Integer> entry : cursors.entrySet()) {
+                if (entry.getValue() < expectedCommitNumber) {
+                    throw new RuntimeException("Node " + entry.getKey() + " has " +
+                        "only reached commit number " + entry.getValue() + ", not " +
+                        expectedCommitNumber);
+                }
+            }
+        }
+    }
+
+    /**
+     * Test that all the log managers see all the commits.
+     */
+    @Test
+    public void testCommits() throws Exception {
+        CommitChecker checker = new CommitChecker(Arrays.asList(
+            new TestCommit(0, 0, new BrokerRecord().setBrokerId(0)),
+            new TestCommit(0, 1, new BrokerRecord().setBrokerId(1)),
+            new TestCommit(0, 2, new BrokerRecord().setBrokerId(2)),
+            new TestCommit(0, 3, new BrokerRecord().setBrokerId(0))
+        ));
+        try (LocalLogManagerTestEnv env = new LocalLogManagerTestEnv(checker, 3)) {
```

Comment:
The number 3 here must match the number 3 below (`checker.verifyCursorsHaveReachedCommitNumber(3, checker.commits.size())`).  Create and share a variable so the values are guaranteed to match?